omniscidb/QueryEngine/Descriptors/QueryFragmentDescriptor.h

```diff
@@ -78,71 +78,61 @@ class QueryFragmentDescriptor {
-  void assignFragsToKernelDispatch(DISPATCH_FCN f,
-                                   const RelAlgExecutionUnit& ra_exe_unit) const {
-    if (execution_kernels_per_device_.empty()) {
-      return;
-    }
-
+  void dispatchKernelsToDevices(DISPATCH_FCN dispatcher_f,
+                                const RelAlgExecutionUnit& ra_exe_unit,
+                                policy::ExecutionPolicy* policy) const {
+    std::unordered_map<ExecutorDeviceType, std::unordered_map<int, size_t>>
+        execution_kernel_index;
     size_t tuple_count = 0;
-
-    std::map<ExecutorDeviceType, std::unordered_map<int, size_t>> execution_kernel_index;
     for (const auto& device_type_itr : execution_kernels_per_device_) {
-      for (const auto& device_itr : device_type_itr.second) {
-        CHECK(execution_kernel_index[device_type_itr.first]
-                  .insert(std::make_pair(device_itr.first, size_t(0)))
-                  .second);
+      if (policy->devices_dispatch_modes.at(device_type_itr.first) ==
+          ExecutorDispatchMode::KernelPerFragment) {
+        for (const auto& device_itr : device_type_itr.second) {
+          CHECK(execution_kernel_index[device_type_itr.first]
+                    .insert(std::make_pair(device_itr.first, size_t(0)))
+                    .second);
+        }
       }
     }
 
-    bool dispatch_finished = false;
-    while (!dispatch_finished) {
-      dispatch_finished = true;
-      for (const auto& device_type_itr : execution_kernels_per_device_)
+    for (const auto& device_type_itr : execution_kernels_per_device_) {
+      if (policy->devices_dispatch_modes.at(device_type_itr.first) ==
+          ExecutorDispatchMode::MultifragmentKernel) {
         for (const auto& device_itr : device_type_itr.second) {
-          auto& kernel_idx =
-              execution_kernel_index[device_type_itr.first][device_itr.first];
-          if (kernel_idx < device_itr.second.size()) {
-            dispatch_finished = false;
-            const auto& execution_kernel = device_itr.second[kernel_idx++];
-            f(device_itr.first,
-              execution_kernel.fragments,
-              rowid_lookup_key_,
-              device_type_itr.first);
-            if (terminateDispatchMaybe(tuple_count, ra_exe_unit, execution_kernel)) {
-              return;
+          const auto& execution_kernels = device_itr.second;
+          const auto& fragments_list = execution_kernels.front().fragments;
+          dispatcher_f(
+              device_itr.first, fragments_list, rowid_lookup_key_, device_type_itr.first);
+        }
+      } else {
+        bool dispatch_finished = false;
+        while (!dispatch_finished) {
+          dispatch_finished = true;
+          for (const auto& device_type_itr : execution_kernels_per_device_)
+            for (const auto& device_itr : device_type_itr.second) {
+              auto& kernel_idx =
+                  execution_kernel_index[device_type_itr.first][device_itr.first];
```

Comment:
Why populate the `execution_kernel_index` at all? Can't you just merge the two loops?

Reply:
Because we are inside the `while (!dispatch_finished){}` loop, so we may actually revisit `for(const auto& device_itr : device_type_itr.second){}`, but to not schedule already scheduled kernels, we keep track of the last scheduled kernel idx to continue from that index on